samples/bluetooth/hci_rpmsg/src/main.c

```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+/* A memory pool debug purposes. Currently used sent HCI vendor specific event with LL_ASSERT
+ * information.
+ */
+NET_BUF_POOL_FIXED_DEFINE(debug_tx_pool, 1, BT_BUF_RX_SIZE, 8, NULL);
+
+static struct net_buf *hci_debug_evt_create(uint8_t evt, uint8_t len)
+{
+	struct bt_hci_evt_hdr *hdr;
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&debug_tx_pool, K_FOREVER);
+	if (buf) {
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = evt;
+		hdr->len = len;
+	}
+
+	return buf;
+}
+
+#endif /* CONFIG_BT_CTLR_ASSERT_HANDLER || CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER */
+
 #if defined(CONFIG_BT_CTLR_ASSERT_HANDLER)
 void bt_ctlr_assert_handle(char *file, uint32_t line)
 {
-	BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
+	uint32_t len = 0U, pos = 0U;
+
+	/* Disable interrupts, this is unrecoverable */
+	(void)irq_lock();
+
+	if (file) {
+		while (file[len] != '\0') {
+			if (file[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file += pos;
+		len -= pos;
+	}
+
+	/* Total length: 0xAA + strlen + \0 + 32-bit line number */
+	len = 1 + len + 1 + 4;
+
+	/* Prepare vendor specific HCI debug event */
+	struct net_buf *buf;
+
+	buf = hci_debug_evt_create(BT_HCI_EVT_VENDOR, len);
+	if (!buf) {
+		BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
+	}
+
+	net_buf_add_u8(buf, 0xAA);
+
+	if (len) {
+		while (*file != '\0') {
+			net_buf_add_u8(buf, *file);
+			file++;
+		}
+		net_buf_add_u8(buf, 0x00);
+	}
+
+	net_buf_add_u8(buf, line >> 0 & 0xff);
+	net_buf_add_u8(buf, line >> 8 & 0xff);
+	net_buf_add_u8(buf, line >> 16 & 0xff);
+	net_buf_add_u8(buf, line >> 24 & 0xff);
+
+	/* Send the event over rpmsg */
+	hci_rpmsg_send(buf);
+
+	while (1) {
+	}
 }
 #endif /* CONFIG_BT_CTLR_ASSERT_HANDLER */
 
+#if defined(CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER)
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* 0xAA + r0/a1 + 0xAA + r1/a2 + 0xAA + r2/a3 + 0xAA + r3/a4 + 0xAA + r12/ip +
+	 * 0xAA + r14/lp + 0xAA + XPSR
+	 */
+	uint32_t len = 7 * (1 + 4);
+
+	/* Disable interrupts, this is unrecoverable */
+	(void)irq_lock();
+
+	if (esf != NULL) {
+		/* Prepare vendor specific HCI debug event */
+		struct net_buf *buf;
+
+		buf = hci_debug_evt_create(BT_HCI_EVT_VENDOR, len);
+		if (!buf) {
+			BT_ASSERT_MSG(false, "Fatal error\n");
+		}
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a1);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a2);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a3);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a4);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.ip);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.lr);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.xpsr);
```

Comment:
```suggestion
		net_buf_add_le32(buf, esf->basic.a1);
		net_buf_add_le32(buf, esf->basic.a2);
		net_buf_add_le32(buf, esf->basic.a3);
		net_buf_add_le32(buf, esf->basic.a4);
		net_buf_add_le32(buf, esf->basic.ip);
		net_buf_add_le32(buf, esf->basic.lr);
		net_buf_add_le32(buf, esf->basic.xpsr);
```

Reply:
This will not give the same, right? 

The original code will be something like
a1, 0xAA, a2, 0xAA, a3, ...
where what you suggest will give
a1, a2, a3

Reply:
Well, I was about to change the implementation to use something that was proposed by @jhedberg:
https://github.com/zephyrproject-rtos/zephyr/blob/360d70a91ea731d47f3a7f84c154f397213f18e6/doc/connectivity/bluetooth/api/hci.txt#L1076
The `fatal_error` is aimed for general fatal errors. It provides PC register and some string with error description e.g. source file and line number. 

Though, it will be an architecture specific, because `z_arch_efs_t` is an alias and its real definition is not common for all architectures. For Cortex-M it will hold: `A1`, `A2`, `A2`, `A3`, `A4`, `IP`, `LR`, XPSR` registers. 

```
#define BT_HCI_EVT_VS_SYS_FATAL_ERROR 0x5
struct bt_hci_evt_vs_sys_fatal_error {
	uint32_t reason;
#if defined(CONFIG_CPU_CORTEX_M)
	uint32_t a1;
	uint32_t a2;
	uint32_t a3;
	uint32_t a4;
	uint32_t ip;
	uint32_t lr;
	uint32_t xpsr;
#endif /* CONFIG_CPU_CORTEX_M */
};
```

I've created a new event, because it seems to me that I can't change the existing one `bt_hci_evt_vs_fatal_error`. It is part of the Zephyrs HCI API that is already released. It will break backward compatibility.

BTW I didn't find any use of `bt_hci_evt_vs_fatal_error` in the Controller. The other thing is that, it is called a Zephyr Fatal Error and description says:
> This event reports a fatal non-recoverable error from the local Controller.

I understand it as a Controller only error. Not system wide, non-recoverable error.

Reply:
@Thalley I was not sure about it, so I wrote my last message to get an input. If there is no problem with change of the `bt_hci_evt_vs_fatal_error` then I'll do that and make it to be system wide fatal error HCI event. To asserts and other Controller only errors I suggest to create a new VS HCI event.

@cvinayak what it your opinion?

Reply:
Well I thought about following distinction:
- `Zephyr Fatal Error` e.g. hard fault, kernel error, etc. anything that is unrecoverable and requires a system/core to be reset
- `Controller Fatal Error` e.g. Controller's LL_ASSERT. That does not mean the core/system has to be restarted, but Controller only. Other errors may fall into that category in future.

For start `Zephyr Fatal Error` would be implemented for Cortex-M. It would include stack frame information like one that is printed by Zephyr in such error happens. The `Controller Fatal Error` would encapsulate information where LL_ASSERT happened: source file and line.

Correct me if I'm wrong in my understanding of the Host view point. Since Host communicates with Controller through a defined HCI interface it does not matter if Controller does not response because of its internal error or system errors. It is just a not responding Controller. 

The thing is a bit different if you take into account conformance tests. While debugging it does matter if the issue is an `LL_ASSERT` or `Zephyr Fatal Error`. Debugging conformance tests is the original reason why I've created these two PRs.

Reply:
Hello all,

Is the current approach okay for you? If it is, then I'll prepare commits to apply same solution for other `hci_XXX1` samples also.